securesystemslib/hsm.py

```diff
@@ -0,0 +1,446 @@
+      securesystemslib.exceptions.NotFoundError, if the path of PKCS#11
+      library is not specified or the library is corrupt
+    """
+
+    # Try to load the PKCS11 library
+    try:
+      # Load the PKCS#11 library and simultaneously update the list
+      # of available HSM.
+      self.PKCS11.load(self.PKCS11LIB)
+    except PyKCS11.PyKCS11Error as error:
+      logger.info('PKS11 Library not found or is corrupt!')
+      raise securesystemslib.exceptions.NotFoundError(error.__str__())
+
+
+
+
+
+  def get_available_HSMs(self):
+    """
+    <Purpose>
+      Generate the list of available cryptographic tokens for the user
+
+    <Exceptions>
+      securesystemslib.exceptions.NotFoundError, if the path of PKCS#11
+      library is not specified or the library is corrupt
+
+    <Returns>
+      A list of dictionaries consisting of relevant information
+      regarding all the available tokens
+    """
+
+    # Refresh the list of available slots for HSM
+    self.refresh()
+
+    # Get the list of slots on which HSMs are available
+    slot_list = self.PKCS11.getSlotList()
+    slot_info_list = []
+
+    # For all the available HSMs available, add relevant information
+    # to the slots dictionary
+    for slot in slot_list:
+      slot_dict = dict()
+      slot_dict['slot_id'] = slot
+      slot_info = self.PKCS11.getSlotInfo(slot)
+      slot_dict['flags'] = slot_info.flags2text()
+      slot_dict['manufacturer_id'] = slot_info.manufacturerID
+      slot_dict['slot_description'] = slot_info.slotDescription
+      slot_info_list.append(slot_dict)
+
+    return slot_info_list
+
+
+
+
+
+  def get_HSM_session(self, slot_info):
+    """
+    <Purpose>
+      Open a session with the HSM of the given 'slot_info'
+
+    <Arguments>
+      slot_info:
+        element from the list returned by get_available_HSMs().
+
+    <Exceptions>
+      securesystemlib.exceptions.InvalidNameError, if the requested token
+      is either not present or cannot be used.
+    """
+    try:
+      self.session = self.PKCS11.openSession(slot_info['slot_id'],
+          PyKCS11.CKF_SERIAL_SESSION | PyKCS11.CKF_RW_SESSION)
+    except PyKCS11.PyKCS11Error as error:
+      raise securesystemslib.exceptions.InvalidNameError(
+          "The requested token is not available." + str(error))
+    except KeyError:
+      raise securesystemslib.exceptions.InvalidNameError(
+          "Invalid Input, not a slot_info dictionary.")
+
+
+
+
+
+  def login(self, user_pin):
+    """
+    <Purpose>
+      User Login into the HSM. Required to access private objects.
+
+    <Arguments>
+      user_pin:
+        PIN for the CKU_USER login.
+
+    <Exceptions>
+      securesystemslib.exceptions.BadPasswordError, if the entered
+      user pin is invalid.
+    """
+
+    try:
+      self.session.login(user_pin)
+    except PyKCS11.PyKCS11Error as error:
+      if error.__str__() == 'CKR_USER_ALREADY_LOGGED_IN (0x00000100)':
+        logger.info('Already logged in as CKU_USER.')
```

Comment:
Nitpick: `logger.warning` might be more semantic.

Also, checking `__str__` is likely to be fragile. It looks like `error.value` contains the numeric error code, which you can map to a symbolic value like `CKR_USER_ALREADY_LOGGED_IN` via `PyKCS11.CKR` dictionary.